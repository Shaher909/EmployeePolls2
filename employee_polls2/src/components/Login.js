import React, { useState } from 'react';
import { connect } from 'react-redux';
import { setAuthUser } from '../actions/authedUser';
import Title from "./Title";
import { useNavigate } from 'react-router-dom'; // Import useNavigate

const Login = ({ users, dispatch }) => {
  const [selectedUser, setSelectedUser] = useState('');
  const [password, setPassword] = useState('');
  const navigate = useNavigate(); // Initialize the navigate hook

  const handleLogin = async () => {
    if (selectedUser && password && users[selectedUser]?.password === password) {
      // Dispatch setAuthUser with the user ID
      dispatch(setAuthUser(selectedUser));

      // Redirect to dashboard after successful login
      navigate('/dashboard');
    } else {
      alert('Invalid username or password');
      // Handle case when no user is selected or password is incorrect
      console.log('Invalid username or password. Please try again.');
    }
  };

  return (
    <div className="formContainer">
      <Title text={"Employees Polls Portal - Login Page"} />
      <label htmlFor="dropdown">Username:</label>
      <br />
      <select
        id="dropdown"
        onChange={(e) => setSelectedUser(e.target.value)}
        value={selectedUser}
      >
        <option value="" disabled>
          Select a username
        </option>
        {Object.keys(users).map((userId) => (
          <option key={userId} value={userId}>
            {users[userId].name}
          </option>
        ))}
      </select>
      <br />
      <label htmlFor="password">Password:</label>
      <br />
      <input
        type="password"
        id="password"
        value={password}
        onChange={(e) => setPassword(e.target.value)}
      />
      <br />
      <input type="button" className="btn" value="Login" onClick={handleLogin} />
    </div>
  );
}

const mapStateToProps = ({ users }) => ({
  users,
});

export default connect(mapStateToProps)(Login);