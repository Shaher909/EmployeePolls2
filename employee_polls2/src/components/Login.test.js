import React from 'react';
import { render, fireEvent, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import Login from './Login';

const mockStore = configureStore();

describe('Login component', () => {
  it('renders username dropdown, password field, and button', () => {
    const users = {
      user1: { id: 'user1', name: 'User 1', password: 'password1' },
      user2: { id: 'user2', name: 'User 2', password: 'password2' },
    };

    const authedUser = null;

    const store = mockStore({
      users,
      authedUser,
    });

    // Render the Login component and destructure asFragment from the result
    const { asFragment, getByLabelText, getByText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    // Use Jest's snapshot feature to create/update snapshots
    // test #5 - snapshot
    expect(asFragment()).toMatchSnapshot();

    //Test #6 DOM Test
    // Get username field and login button
    const usernameField = getByLabelText('Username:');
    const loginButton = getByText('Login');

    if (usernameField && loginButton) {
      // Trigger a change event in the username field
      fireEvent.change(usernameField, { target: { value: 'user1' } });

      // Verify that the username field is updated
      expect(usernameField).toHaveValue('user1');

      // Trigger a click event on the login button
      fireEvent.click(loginButton);

    } else {
      // Handle the case where elements are not found
      throw new Error('Login button or username field not found');
    }
  });
});