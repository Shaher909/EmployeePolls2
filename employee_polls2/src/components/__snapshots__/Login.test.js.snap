// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders username dropdown, password field, and button 1`] = `
<DocumentFragment>
  <div
    class="formContainer"
  >
    <div>
      <h1>
         Employees Polls Portal - Login Page 
      </h1>
    </div>
    <label
      for="dropdown"
    >
      Username:
    </label>
    <br />
    <select
      id="dropdown"
    >
      <option
        disabled=""
        value=""
      >
        Select a username
      </option>
      <option
        value="user1"
      >
        User 1
      </option>
      <option
        value="user2"
      >
        User 2
      </option>
    </select>
    <br />
    <label
      for="password"
    >
      Password:
    </label>
    <br />
    <input
      id="password"
      type="password"
      value=""
    />
    <br />
    <input
      class="btn"
      type="button"
      value="Login"
    />
  </div>
</DocumentFragment>
`;
